pkg/minishift/update/update.go

```diff
@@ -0,0 +1,336 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package update
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/google/go-github/github"
+	update "github.com/inconshreveable/go-update"
+	"github.com/minishift/minishift/pkg/util/archive"
+	pb "gopkg.in/cheggaaa/pb.v1"
+
+	githubutils "github.com/minishift/minishift/pkg/util/github"
+	"github.com/minishift/minishift/pkg/version"
+)
+
+const (
+	timeLayout  = time.RFC1123
+	githubOwner = "minishift"
+	githubRepo  = "minishift"
+)
+
+// CurrentVersion returns the current version of minishift binary installed on the system
+func CurrentVersion() (semver.Version, error) {
+	localVersion, err := version.GetSemverVersion()
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return localVersion, nil
+}
+
+// LatestVersion returns the latest version of minishift binary available from upstream
+func LatestVersion() (semver.Version, error) {
+	latestVersion, err := getLatestVersionFromGitHub(githubOwner, githubRepo)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return latestVersion, nil
+}
+
+// IsNewerVersion compares the local and latest versions and returns a boolean
+func IsNewerVersion(localVersion, latestVersion semver.Version) bool {
+	if localVersion.Compare(latestVersion) < 0 {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Update handles the update process by calling functions that download, verify, extract and replace the binary.
+// It returns an error if any of these functions fails at any point.
+func Update(output io.Writer, latestVersion semver.Version) error {
+
+	var extName string
+
+	if runtime.GOOS == "windows" {
+		extName = "zip"
+	} else {
+		extName = "tgz"
+	}
+
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", latestVersion, runtime.GOOS, runtime.GOARCH, extName)
+	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+
+	fmt.Printf("Updating to version %s\n", latestVersion)
+
+	url := fmt.Sprintf(downloadLinkFormat, latestVersion, archiveName)
+	// downloadedAchive is a string that has the path to archive on the disk
+	downloadedArchive, err := downloadAndVerifyArchive(url)
+	if err != nil {
+		return err
+	}
+
+	dir := strings.Split(downloadedArchive, "minishift")[0]
+
+	defer os.RemoveAll(dir)
+
+	// Extract and replace the downloaded archive in place of current binary
+
+	binary, err := extractBinary(downloadedArchive)
+	if err != nil {
+		return err
+	}
+
+	err = updateBinary(binary)
+	if err != nil {
+		return err
+	}
+
+	// env := os.Environ()
```

Comment:
We're not verifying it by the way of executing `minishift version` because that doesn't work well on Windows. If you have suggestion(s) around some other way of doing it, I'm open to doing it.